app/services/bulk_suite_selector.rb

```diff
@@ -0,0 +1,136 @@
+#
+# Service object to handle the bulk assignment of suites to groups. Ensures that
+# suites are available, groups don't have suites assigned, and optionally that
+# suites belong to the appropriate draw.
+class BulkSuiteSelector
+  include ActiveModel::Model
+
+  validate :all_groups_selected
+  validate :no_duplicate_selections
+  validate :suites_in_correct_draw
+
+  # Initialize a new BulkSuiteSelector.
+  #
+  # @param groups [Array<Group>] the groups to assign suites to
+  def initialize(groups:)
+    @groups = groups
+  end
+
+  # Prepare the BulkSuiteSelector by processing form submission params
+  #
+  # @param params [#to_h] the controller params
+  # @return [BulkSuiteSelector] the service object
+  def prepare(params:)
+    @params = params.to_h.transform_keys(&:to_sym)
+    @params.each do |k, v|
+      next unless respond_to_missing?(k) && !v.empty?
+      instance_variable_set("@#{k}", v)
+    end
+    self
+  end
+
+  # Process the suite selections
+  #
+  # @return [Hash{Symbol=>Nil,BulkSuiteSelector,Hash}] the results hash
+  def process
+    return error(errors.full_messages.join(', ')) unless valid?
+    ActiveRecord::Base.transaction do
+      process_all_suite_selections
+    end
+    success
+  rescue ActiveRecord::ActiveRecordError => e
+    error(e.to_s)
+  end
+
+  # Returns the valid form field ids for a given set of groups
+  #
+  # @return [Array<Symbol>] the list of form fields
+  def valid_field_ids
+    return [] unless groups
+    @valid_field_ids ||= groups.map { |g| param_for(g) }
+  end
+
+  private
+
+  attr_reader :groups, :params
+
+  # note that this occurs in a transaction
+  def process_all_suite_selections
+    results = groups.each_with_object({}) do |g, hash|
+      hash[g] = process_suite_selection(g)
+    end
+    return unless results.values.any? { |r| !r.empty? }
+    raise ActiveRecord::ActiveRecordError, errors_from_results(results)
+  end
+
+  def process_suite_selection(g)
+    suite_id = send(param_for(g))
+    selector = SuiteSelector.new(group: g, suite_id: suite_id)
+    selector.select
+    selector.errors
+  end
+
+  def all_groups_selected
+    return if all_groups_selected?
+    errors.add(:base, 'You must select a suite for all groups')
+  end
+
+  def all_groups_selected?
+    params.select { |_k, v| !v.empty? }.count == groups.count
+  end
+
+  def no_duplicate_selections
+    return if params.values.uniq == params.values
+    errors.add(:base, 'You must select different suites for each group')
+  end
+
+  def suites_in_correct_draw
+    return unless all_groups_selected?
+    return unless groups.any? do |g|
+      next false unless g.draw
+      suite = Suite.find_by(id: send(param_for(g)))
+      next false if suite&.draws&.include? g.draw
+      true
+    end
+    errors.add(:base, 'All groups in draws must be assigned to suites in '\
+      'the same draw')
+  end
+
+  def errors_from_results(results_hash)
+    results_hash.each_with_object([]) do |(g, r), array|
```

Comment:
These errors are kinda reachable from the front-end, and I'd rather be more verbose now and deal with it down the road when we figure out the frontend properly. I renamed the block variables for clarity.